src/Streamly/Internal/Data/Stream/Async/Channel/Dispatcher.hs

```diff
@@ -62,65 +62,102 @@ pushWorker yieldMax sv = do
+    -- processed before the addThread occurs, so in that case we have to use
+    -- modifyThread which performs a toggle rather than adding or deleting.
+    --
+    -- XXX We can use addThread or modThread based on eager flag.
+    doFork (workLoop sv winfo) (svarMrun sv) exception >>= modThread
 
     where
 
+    modThread = modifyThread (workerThreads sv) (outputDoorBell sv)
     exception = handleChildException (outputQueue sv) (outputDoorBell sv)
 
--- Returns:
--- True: can dispatch more
--- False: cannot dispatch any more
-dispatchWorker :: (MonadIO m, MonadBaseControl IO m) => Count -> Channel m a -> m Bool
-dispatchWorker yieldCount sv = do
+{-# INLINE getEffectiveWorkerLimit #-}
+getEffectiveWorkerLimit :: MonadIO m => Channel m a -> m Limit
+getEffectiveWorkerLimit sv = do
     let workerLimit = maxWorkerLimit sv
+    case remainingWork sv of
+        Nothing -> return workerLimit
+        Just ref -> do
+            n <- liftIO $ readIORef ref
+            case yieldRateInfo sv of
+                Just _ -> return workerLimit
+                Nothing ->
+                    return $
+                        case workerLimit of
+                            Unlimited -> Limited (fromIntegral n)
+                            Limited lim -> Limited $ min lim (fromIntegral n)
+
+{-# INLINE checkMaxThreads #-}
+checkMaxThreads :: MonadIO m => Int -> Channel m a -> m Bool
+checkMaxThreads active sv = do
+    -- Note that we may deadlock if the previous workers (tasks in the
+    -- stream) wait/depend on the future workers (tasks in the stream)
+    -- executing. In that case we should either configure the maxWorker
+    -- count to higher or use parallel style instead of ahead or async
+    -- style.
+    limit <- getEffectiveWorkerLimit sv
+    return
+        $ case limit of
+            Unlimited -> True
+            -- Note that the use of remainingWork and workerCount is not
+            -- atomic and the counts may even have changed between reading
+            -- and using them here, so this is just approximate logic and
+            -- we cannot rely on it for correctness. We may actually
+            -- dispatch more workers than required.
+            Limited lim -> fromIntegral lim > active
+
+{-# INLINE checkMaxBuffer #-}
+checkMaxBuffer :: MonadIO m => Int -> Channel m a -> m Bool
+checkMaxBuffer active sv = do
+    let limit = maxBufferLimit sv
+    case limit of
+        Unlimited -> return True
+        Limited lim -> do
+            (_, n) <- liftIO $ readIORef (outputQueue sv)
+            return $ fromIntegral lim > n + active
+
+dispatchWorker :: (MonadIO m, MonadBaseControl IO m) =>
+    Count -> Channel m a -> m Bool
+dispatchWorker yieldCount sv = do
     -- XXX in case of Ahead streams we should not send more than one worker
     -- when the work queue is done but heap is not done.
+    -- XXX Should we have a single abstraction for checking q and
```

Comment:
q is work right?

Reply:
In case of ahead we have a queue and a heap, both together are work. The queue may be empty but the heap may have some pending work to do.